controllers/build.go

```diff
@@ -0,0 +1,263 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"encoding/json"
+	"strings"
+
+	routev1 "github.com/openshift/api/route/v1"
+	appstudiov1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
+	tektonapi "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	triggersapi "github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func determineBuildDefinition(component appstudiov1alpha1.Component) string {
+	return "devfile-build"
+}
+
+func determineBuildCatalog(namespace string) string {
+	// TODO: If there's a namespace/workspace specific catalog, we got
+	// to respect that.
+	return "quay.io/redhat-appstudio/build-templates-bundle:v0.1.2"
+}
+
+// determineBuildExecution returns the pipelineRun spec
+func determineBuildExecution(component appstudiov1alpha1.Component, params []tektonapi.Param, workspaceSubPath string) tektonapi.PipelineRunSpec {
+
+	pipelineRunSpec := tektonapi.PipelineRunSpec{
+		Params: params, //paramsForWebhookBasedBuilds(component),
+		PipelineRef: &tektonapi.PipelineRef{
+
+			// This can't be hardcoded to devfile-build.
+			// The logic should determine if it is a
+			// nodejs build, java build, dockerfile build or a devfile build.
+			Name:   determineBuildDefinition(component),
+			Bundle: determineBuildCatalog(component.Namespace),
+		},
+
+		Workspaces: []tektonapi.WorkspaceBinding{
+			{
+				Name: "workspace",
+				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+					ClaimName: "appstudio",
+				},
+				SubPath: component.Name + "/" + workspaceSubPath, //$(tt.params.git-revision)/",
+			},
+			{
+				Name: "registry-auth",
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: "redhat-appstudio-registry",
+				},
+			},
+		},
+	}
+	return pipelineRunSpec
+}
+
+func normalizeOutputImageURL(outputImage string) string {
+
+	// if provided image format was
+	//
+	// quay.io/foo/bar:mytag , then
+	// push to quay.io/foo/bar:mytag-git-revision
+	//
+	// else,
+	// push to quay.io/foo/bar:git-revision
+
+	if strings.Contains(outputImage, ":") {
+		outputImage = outputImage + "-" + "$(tt.params.git-revision)"
+	} else {
+		outputImage = outputImage + ":" + "$(tt.params.git-revision)"
+	}
+	return outputImage
+}
+
+func paramsForInitialBuild(component appstudiov1alpha1.Component) []tektonapi.Param {
```

Comment:
would you mind pls putting a small desc what the func does, so that i dont have to scroll and check for a quick glance every time i come across it

probably a good idea for all the funcs in the file